python/kfserving/kfserving/storage.py

```diff
@@ -43,15 +47,48 @@ def download(uri: str) -> str:
 
     @staticmethod
     def _download_s3(uri, temp_dir: str):
-        raise NotImplementedError
+        client = Storage._create_minio_client()
+        bucket_args = uri.replace(_S3_PREFIX, "", 1).split("/", 1)
+        bucket_name = bucket_args[0]
+        bucket_path = bucket_args[1] if len(bucket_args) > 1 else ""
+        objects = client.list_objects(bucket_name, prefix=bucket_path, recursive=True)
+        for obj in objects:
+            object_file_name = obj.object_name.replace(bucket_path, "", 1).strip("/")
```

Comment:
How about basename? https://docs.python.org/2/library/os.path.html

Reply:
Maybe the name here is confusing, I renamed the variable as `subdir_object_key`. What we want to do over here is to replace the prefix from s3 or gcs with the temp_dir path on the local file system. I didn't use `basename` since we want to keep the object key with all the subdirectories after the prefix and append it to `temp_dir`.

Reply:
Makes sense.